### ISSUE TITLE
Plugin fails to set Query and Headers in C++

### ISSUE THREAD

**Issue by TheFurrow** [getnamo/socketio-client-ue4#70]
Hi!

I am not sure is this issue or it just does not work like I figured out. So I am trying to set Query and Headers in C++ but without results, they work fine in Blueprint. I am using ConnectNative-function (the one eats TSharedPtr<FJsonObject>).

Here is my code for later inspection

```c++
//CharacterBase.cpp

ACharacterBase::ACharacterBase(const FObjectInitializer& ObjectInitializer)
{
        /* Some unrelated stuff above here */

	SocketIO = CreateDefaultSubobject<USocketIOClientComponent>(TEXT("SocketIOClient"));
	SocketIO->bAutoActivate = false;
	SocketIO->bShouldAutoConnect = false;
	SocketIO->SetIsReplicated(false);
	SocketIO->AddressAndPort = FString("http://127.0.0.1:8080");
}
```

```c++
//PlayerCharacter.cpp

SocketIO->bShouldAutoConnect = false;

TSharedPtr<FJsonObject> Query = MakeShareable(new FJsonObject);
Query->SetStringField(FString("auth"), FString("bbbbbbb"));

SocketIO->ConnectNative(SocketIO->AddressAndPort, Query);
```

```js
//server.js
playersNSP.use(function(socket, next) {
	if(socket != undefined)
	{
		console.log(socket); //Print below
		if(socket.handshake.query.auth == serverConfig.secret) {
			console.log('Authorized connection for %s', socket.id);
			next();
			return;
		}
		
		console.log('Unauthorized connection');
		return;
	}
	else {
		console.log('Unauthorized connection');

		return;
	}
});
```

```text
//output result of console.log(socket)
...
handshake:
 { headers:
    { connection: 'Upgrade',
      host: '127.0.0.1:8080',
      'sec-websocket-key': '7e6NiRKQznANwKqrtu5+GA==',
      'sec-websocket-version': '13',
      upgrade: 'websocket',
      'user-agent': 'WebSocket++/0.7.0' },
   time: 'Mon Jan 22 2018 14:02:21 GMT+0200 (FLE Standard Time)',
   address: '127.0.0.1',
   xdomain: false,
   secure: false,
   issued: 1516622541842,
   url: '/socket.io/?EIO=4&transport=websocket&t=1516622541',
   query: { EIO: '4', transport: 'websocket', t: '1516622541' } },
```

And the very same with Blueprint

![Blueprint construction phase](https://image.ibb.co/h9xAww/Untitled.png)

```text
....
  handshake:
   { headers:
      { connection: 'Upgrade',
        host: '127.0.0.1:8080',
        'sec-websocket-key': 'G+CcGU6PYtQFB9ao1ADSdA==',
        'sec-websocket-version': '13',
        upgrade: 'websocket',
        'user-agent': 'WebSocket++/0.7.0' },
     time: 'Mon Jan 22 2018 14:13:23 GMT+0200 (FLE Standard Time)',
     address: '127.0.0.1',
     xdomain: false,
     secure: false,
     issued: 1516623203369,
     url: '/socket.io/?EIO=4&transport=websocket&t=1516623203&auth=bbbbbbb',
     query:
      { EIO: '4',
        transport: 'websocket',
        t: '1516623203',
        auth: 'bbbbbbb' } },
....
```

I tried to inspect the problem by myself, without any results. As far as I could go, ended up in SocketIONative.cpp, which where everything were fine (I think)

```c++
//SocketIONative.cpp
PrivateClient->connect(StdAddressString, QueryMap, HeadersMap);
```

Just one notify to releaser, Getnamo, yes I tried newer version of plugin with failure to compile.

### Versions I am using
UE Plugin version: 0.6.3 
Unreal Engine version: 4.16
Socket.IO version: 2.0.4

**Comment by getnamo** [comment]
Out of curiosity if you use the blueprint function in c++ or tried to connect with both Query and Headers? e.g. make an extra empty FJsonObject

```TSharedPtr<FJsonObject> Headers = MakeShareable(new FJsonObject);```

and pass that in as

```SocketIO->ConnectNative(SocketIO->AddressAndPort, Query, Headers);```

I suspect having either null may turn both of them off.

**Comment by TheFurrow** [comment]
No luck with both Query and Headers set. Made two testes, on first test Headers was empty, no data included, while Query had data in it and vice versa.

After making some changes to the SocketIOClientComponent.cpp, changed

```c++
{
NativeClient->Connect(InAddressAndPort, Query, Headers);
}
```

**Comment by getnamo** [comment]
It's a bit strange behavior. It will likely need the section from https://github.com/getnamo/socketio-client-ue4/blob/master/Source/SocketIOClient/Private/SocketIOClientComponent.cpp#L348 to https://github.com/getnamo/socketio-client-ue4/blob/master/Source/SocketIOClient/Private/SocketIOClientComponent.cpp#L364 moved to connectnative.

**Comment by getnamo** [comment]
check out the Issue-70 branch https://github.com/getnamo/socketio-client-ue4/tree/issue-70 with these changes https://github.com/getnamo/socketio-client-ue4/commit/79ce01484cba34b081f8d06689c4f9a7c36834cc

Let me know if that fixes connect native.

**Comment by TheFurrow** [comment]
No luck with those changes. Works only if user have defined the FJsonObject _and it is pointing to an object_ and they are validated before passing them in `NativeClient->Connect()`.

**Comment by getnamo** [comment]
Whoops you're right, forgot to initialize them in the invalid case. Added https://github.com/getnamo/socketio-client-ue4/commit/832fbc2489b09ce452e2bf5d69a3f7dac38a67a9

re-pull the 70 branch and let me know if that fixed it.

**Comment by TheFurrow** [comment]
Not at all. I checked them with `IsValid()` after setting _objects_ having result of `true`

**Comment by SirakDan** [comment]
Any news about this?
I'm following a compilling error on Linux and it seems to head down to that specific point in the code, the:
`PrivateClient->connect(StdAddressString, QueryMap, HeadersMap);`

If it helps, it says that the arguments are wrong. I've checked the output and in the Query and Headers arguments there are a couple of  ">" at the end that would make it crash. 
Specifically, at the QueryMap argument there is one extra ">" and in the HeadersMap there are two extras ">" (closing half-diamond).

**Comment by getnamo** [comment]
it looks like it's having a problem with https://github.com/getnamo/socketio-client-ue4/blob/master/Source/SocketIOClient/Private/SocketIONative.cpp#L51 which likely points to the linux compiler having issues with default references to smart pointers.

Out of curiosity have you included https://github.com/getnamo/socketio-client-ue4/compare/issue-70 changes for your compile tests?

Otherwise you may wish to change https://github.com/getnamo/socketio-client-ue4/blob/master/Source/SocketIOClient/Public/SocketIONative.h#L88 e.g 
from references 
```c++
void Connect(	
		const FString& InAddressAndPort,
		const TSharedPtr<FJsonObject>& Query /*= nullptr*/, 
		const TSharedPtr<FJsonObject>& Headers /*= nullptr*/);
```

to just
```c++
void Connect(	
		const FString& InAddressAndPort,
		TSharedPtr<FJsonObject> Query, 
		TSharedPtr<FJsonObject> Headers);
```

There may be some method calls with null references youll just need to add empty objects, but if this works we will likely just need to change the connect function with overloads for linux compile to work.

**Comment by SirakDan** [comment]
Hi getnamo!
I'm afraid none of those solutions worked for me. I included the issue-70 code, changed the Connect header. I've tried to initialize every object to an empty one meeting the same fate.

Is there any other stuff or direction you recommend me heading to?

Thank you!

**Comment by getnamo** [comment]
Hmm, I think I'd need to see the debug stack output myself to figure this one out. Could you provide me with details on how you're crosscompiling to linux so I can repro?

**Comment by SirakDan** [comment]
You are awesome!
I'm compiling this with UnrealEngine 4.17.2-release (the source code), following the official wiki.
I'm using cross-compiling libraries, exactly: v9_clang-4.0.0-centos7 as told in the wiki article:
https://wiki.unrealengine.com/Compiling_For_Linux

Then, I just copy your source code into the plugin directory, execute the "Development Server" > "Linux" configuration when compiling and then I get the error I've shown in my first post.

Is there anything else you need?

Thanks!

**Comment by SirakDan** [comment]
Hi again, 
I can confirm this issue with the Linux compilation is also happening with UnrealEngine 4.18.3 and clang v10.
I'm currently trying to rebuild the dependencies libs without much luck to see if I can bypass the issue.

Greetings and thanks!

**Comment by getnamo** [comment]
Try https://github.com/getnamo/socketio-client-ue4/releases/tag/0.9.0 release and let me know if issue has been fixed or not
